### DISCUSSION TITLE
Limitations / issues of `expect-type`

### DISCUSSION THREAD

**Discussion by mrazauskas** [true-myth/true-myth#862] [General]
This note is inspired by `semver-ts`. It is very useful resource and I can only agree with the ideas and approaches expressed there. Although one detail remains questionable.

As an author of a type testing tool I am obviously biased, but it felt like `expect-type` was overestimated. Take a look at the following examples ([playground](https://www.typescriptlang.org/play/?ts=5.7.2#code/JYWwDg9gTgLgBAbzgWQIYE8BGBTANHAOQhgAtgA7AczgF84AzKCEOAchigFdsBaEdUgHoQGHKwDcAKFCRYiONgAeYbAGMYAFXQqA8vVoMmLVkpXqeMbdgmTJqiOQDO8csTJUd5LSrgBeFKLYAHSupBSUADzOUOEAfAAUAJRSpmqaVnrxoe6Unt7YcKiOheToiUEwEGgwqiT5ehFoWNhRHHEJyXCCgnBgRY4AhEO2ADbY8CLNfgHNQRD08awk2CMjEKzJkqnq9QuTOIXF5NgAbthQ5ZUAEqhnAApMKrDoi8CORGFUG+JdPX2Og2Gkm6cAAAjBHDxtjAoVAmFAAFxwB4QJ6WNhvD45VhwAAmEGwR2ICkUb3gDjglh8rGOZygrEk8X2BSKcFp53KmLc4SkkjG8CUqHAY2mSSRSFQAH4kdFwnAAD5wTjkXHYegUbC4gy+WJweIIGibLbKNK7eJNHBzBaC4XYRKXCAAUQAjpxUCNdo1AhEkn5dRLpXBZVRaLEOj8Qf9AQNjWZ0roFhbgvN4jawGN7RUna73Z6kz7En75FKZW0Q4rlar1cctTQw0kpEA)):

```ts
import { Maybe, Nothing } from 'true-myth/maybe';
import { expectTypeOf } from 'expect-type';

const nothingOnType = Maybe.nothing<string>();
expectTypeOf(nothingOnType as any).toMatchTypeOf<Maybe<string>>(); // pass!!!

let maybe = Maybe.of('hello');
expectTypeOf(maybe as never).toHaveProperty('isNothing'); // pass!!!

// @ts-expect-error: Property 'isNothing' does not exist on type 'never'
(maybe as never).isNothing;

let example = (): { a?: string | undefined } => ({});

expectTypeOf(Maybe.of(example)).toEqualTypeOf<Maybe<() => { a?: string }>>(); // pass!!!
expectTypeOf(Maybe.of(example)).toEqualTypeOf<Maybe<() => { a?: string | undefined }>>();
```

These assertions are taken/adapted from `true-myth` repo. As you see, `.toMatchTypeOf()` and `.toHaveProperty()` do not handle `any` and `never`. Also the `exactOptionalPropertyTypes` option is not taken into account by `.toEqualTypeOf()`.

These are the limitations of `expect-type`s implementation. For instance, `keyof never` resolves to `any` and that is the problem behind `.toHaveProperty()`. (In my opinion, correctness is the major problem of the assertions that are implemented as generic types. In contrary, comparing types programatically is an advantage, because this is how TypeScript compares them.)

`true-myth` is mentioned as one of adopters of `semver-ts`. It would be interesting to hear your opinion.

**Comment by mrazauskas** [comment]
Hm.. This may look like a discussion, but I opened it as an issue on purpose.

For instance, `semver-ts` specification states that `exactOptionalPropertyTypes: true` must be set. The recommendation cannot be enforced, if the option is ignored. Based on this I make a conclusion that the tool is not `semver-ts` compliant. Or?

The recommendation has really good reasons. It prevents bugs. So I felt like this limitation is a call to action.

**Reply by chriskrycho** [reply]
I appreciate your enthusiasm, but choosing to use one type testing library over another is absolutely not a *bug* or an *issue*—per the SemVer TS spec (remember: I *wrote* that spec!) or otherwise. It’s a discussion because I think it’s interesting and worth exploring.

To be direct, you’re bordering on antagonistic in the way you’re pushing TSTyche. I am glad you’ve spent the time to build it, and I remain interested in exploring its tradeoffs. However, treating *not* using it as a problem, or acting as if using long-standing and well-regarded libraries in the space is a problem, will not earn you any friends.

As I said in my comment on the SemVer TS repo, I am interested to spend some time with TSTyche as a tool and evaluate it as something I may add as an *additional* recommendation. I will, however, be doing that on my own timeline, not anyone else’s. Also, bluntly, I will absolutely *not* be recommending it if you continue to be pushy about it.

Please do keep up the good work in the space; please also be patient and respectful of others’ time and expertise. Thanks!
